examples/aggregator/module/rust/src/lib.rs

```diff
@@ -22,76 +22,117 @@
+use proto::aggregator::{Sample, SerializedSparseVector};
+use proto::aggregator_grpc::{Aggregator, AggregatorClient, Dispatcher};
 use protobuf::well_known_types::Empty;
+use std::collections::HashMap;
+use std::convert::{From, TryFrom};
 
-const SAMPLE_THRESHOLD: u64 = 3;
+/// Currently threshold value is hardcoded.
+const SAMPLE_THRESHOLD: u64 = 2;
 
-oak::entrypoint!(oak_main => {
-    oak::logger::init_default();
-    let node = AggregatorNode {
-        aggregator: ThresholdAggregator::<Vector>::new(SAMPLE_THRESHOLD),
-    };
-    Dispatcher::new(node)
-});
+type ThresholdAggregatorMap = HashMap<String, Option<ThresholdAggregator<SparseVector>>>;
 
-impl Monoid for Vector {
-    fn identity() -> Self {
-        Vector::new()
-    }
+/// Oak Node that collects aggregated data.
+pub struct AggregatorNode {
+    aggregators: ThresholdAggregatorMap,
+}
 
-    /// Adds two non-negative integer vectors elementwise. If vectors have different lengths, then
-    /// pads the smallest vector with zeros.
-    fn combine(&self, other: &Self) -> Self {
-        use itertools::EitherOrBoth::*;
-        use itertools::Itertools;
+impl AggregatorNode {
+    fn new() -> Self {
+        AggregatorNode {
+            aggregators: ThresholdAggregatorMap::new(),
+        }
+    }
 
-        Vector {
-            items: self
-                .items
-                .iter()
-                .zip_longest(other.items.iter())
-                .map(|p| match p {
-                    Both(l, r) => *l + *r,
-                    Left(l) => *l,
-                    Right(r) => *r,
-                })
-                .collect(),
-            ..Default::default()
+    fn submit_sparse_vector(&mut self, bucket: String, svec: &SparseVector) -> Result<(), String> {
+        match self.aggregators.get_mut(&bucket) {
+            Some(entry) => match *entry {
+                Some(ref mut aggregator) => {
+                    aggregator.submit(svec);
+                    if let Some(aggregated_data) = aggregator.take() {
+                        *entry = None;
+                        self.send_aggregated_data(bucket, aggregated_data);
+                    }
+                }
+                None => Err(format!("Outdated bucket: {}", bucket))?,
+            },
+            None => {
+                let mut aggregator = ThresholdAggregator::<SparseVector>::new(SAMPLE_THRESHOLD);
+                aggregator.submit(svec);
+                self.aggregators.insert(bucket, Some(aggregator));
+            }
         }
+        Ok(())
     }
-}
 
-/// Oak Node that collects aggregated data.
-pub struct AggregatorNode {
-    aggregator: ThresholdAggregator<Vector>,
+    fn send_aggregated_data(&self, bucket: String, svec: SparseVector) {
+        debug!(
+            "Sending aggregated data: bucket {}, sparse vector: {:?}",
+            bucket, svec
+        );
+
+        match oak::grpc::client::Client::new("grpc-client", "ignored").map(AggregatorClient) {
```

Comment:
What's `"ignored"`?

Reply:
I'm still not sure what this is used for. How can a gRPC pseudo-node have an entrypoint name? Perhaps @daviddrysdale can clarify.

Reply:
A pseudo-Node doesn't have an entrypoint name, but the same `node_create` host function is used for creating both Wasm Nodes (which need a module/config name and entrypoint name) and pseudo-Nodes (which just need a config name); the `"ignored"` placeholder is a reminder.

Reply:
Indeed, so why expose the parameter in `Client::new` at all? AFAICT no one would even pass anything useful via it. It will still use `node_create` internally of course, but the placeholder will be hidden inside the implementation of `new`, where it can also be documented.

Reply:
It's needed because another Wasm node may provide a gRPC service, and the `grpc::client:Client` code works with either a remote (non-Oak) gRPC service (via the gRPC client pseudo-Node), or a local gRPC service (provided by another Oak Node in the same Application).  

TL;DR : [It's required here](https://github.com/project-oak/oak/blob/master/examples/hello_world/module/rust/src/lib.rs#L29)